File: src/lib/workspace/commands.ts
/**
 * Typed wrappers around the Rust workspace commands.
 *
 * Anything that can lose data lives in Rust: directory creation, renames,
 * moves, and deletes. The backend holds the approved workspace root and only
 * ever accepts relative paths from here, so a bad value cannot redirect an
 * operation at a folder outside the workspace.
 */
import { invoke } from '@tauri-apps/api/core';

export interface ApprovedWorkspaceInfo {
  /** Absolute, symlink-resolved path to the workspace root. */
  path: string;
  /** Final path segment, used as the display name. */
  name: string;
  /** Absolute path to the `.notara` sidecar directory. */
  sidecarPath: string;
}

export interface DeletionPreview {
  isDirectory: boolean;
  fileCount: number;
  directoryCount: number;
}

/**
 * Approves a folder as the workspace root.
 *
 * This widens the desktop filesystem scope to cover the folder and creates
 * `.notara` with its `backups` and `media` subdirectories. It is safe to call
 * again with the same path, which is how the app reconnects after a restart.
 */
export const approveWorkspace = (path: string): Promise<ApprovedWorkspaceInfo> =>
  invoke<ApprovedWorkspaceInfo>('approve_workspace', { path });

export const forgetWorkspace = (): Promise<void> => invoke<void>('forget_workspace');

export const createWorkspaceDirectory = (relativePath: string): Promise<string> =>
  invoke<string>('create_workspace_directory', { relativePath });

export const renameWorkspaceEntry = (relativePath: string, newName: string): Promise<string> =>
  invoke<string>('rename_workspace_entry', { relativePath, newName });

export const moveWorkspaceEntry = (
  relativePath: string,
  destinationDirectory: string
): Promise<string> =>
  invoke<string>('move_workspace_entry', { relativePath, destinationDirectory });

export const deleteWorkspaceEntry = (relativePath: string): Promise<void> =>
  invoke<void>('delete_workspace_entry', { relativePath });

export const previewWorkspaceDeletion = (relativePath: string): Promise<DeletionPreview> =>
  invoke<DeletionPreview>('preview_workspace_deletion', { relativePath });

File: src-tauri/src/workspace/commands.rs
//! Tauri commands for workspace approval and directory work.
//!
//! Every command here is a thin wrapper. It resolves the approved root from
//! backend state and hands the work to `dirs`, which holds the logic and the
//! tests. Relative paths are the only thing the webview gets to supply, so a
//! bad value cannot redirect an operation at a folder the user never approved.

use std::path::Path;

use serde::Serialize;
use tauri::{AppHandle, Runtime, State};
use tauri_plugin_fs::FsExt;

use super::dirs::{self, DeletionPreview};
use super::guard::{canonical_root, SIDECAR_DIRECTORY};
use super::state::ApprovedWorkspace;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ApprovedWorkspaceInfo {
    /// Absolute, symlink-resolved path to the workspace root.
    pub path: String,
    /// Final path segment, used as the display name.
    pub name: String,
    /// Absolute path to the `.notara` sidecar directory.
    pub sidecar_path: String,
}

fn display_name(root: &Path) -> String {
    root.file_name()
        .map(|name| name.to_string_lossy().to_string())
        .unwrap_or_else(|| root.to_string_lossy().to_string())
}

/// Approves a folder as the workspace root.
///
/// Three things happen here and the order matters. The path is resolved and
/// checked first, then the filesystem scope is widened to cover it, then the
/// sidecar directories are created. Widening the scope before the path is known
/// good would hand out access based on an unverified string.
///
/// The sidecar gets its own scope entry. On Unix the plugin's glob matching
/// requires a literal leading dot, so a recursive rule on the root alone never
/// matches anything inside `.notara`.
#[tauri::command]
pub fn approve_workspace<R: Runtime>(
    app: AppHandle<R>,
    workspace: State<'_, ApprovedWorkspace>,
    path: String,
) -> Result<ApprovedWorkspaceInfo, String> {
    let root = canonical_root(Path::new(&path))?;
    let sidecar = root.join(SIDECAR_DIRECTORY);

    let scope = app.fs_scope();
    scope
        .allow_directory(&root, true)
        .map_err(|error| format!("Unable to grant access to {}: {error}", root.display()))?;
    scope
        .allow_directory(&sidecar, true)
        .map_err(|error| format!("Unable to grant access to {}: {error}", sidecar.display()))?;

    for directory in [sidecar.clone(), sidecar.join("backups"), sidecar.join("media")] {
        std::fs::create_dir_all(&directory)
            .map_err(|error| format!("Unable to create {}: {error}", directory.display()))?;
    }

    workspace.set(root.clone());

    Ok(ApprovedWorkspaceInfo {
        name: display_name(&root),
        path: root.to_string_lossy().to_string(),
        sidecar_path: sidecar.to_string_lossy().to_string(),
    })
}

/// Drops the approved root so later commands fail closed.
///
/// The filesystem scope is not narrowed here. The plugin has no revoke call,
/// and reconnecting to the same folder is the common case, so leaving the scope
/// alone avoids a permission prompt the user already answered.
#[tauri::command]
pub fn forget_workspace(workspace: State<'_, ApprovedWorkspace>) -> Result<(), String> {
    workspace.clear();
    Ok(())
}

#[tauri::command]
pub fn create_workspace_directory(
    workspace: State<'_, ApprovedWorkspace>,
    relative_path: String,
) -> Result<String, String> {
    dirs::create_directory(&workspace.require()?, &relative_path)
}

#[tauri::command]
pub fn rename_workspace_entry(
    workspace: State<'_, ApprovedWorkspace>,
    relative_path: String,
    new_name: String,
) -> Result<String, String> {
    dirs::rename_entry(&workspace.require()?, &relative_path, &new_name)
}

#[tauri::command]
pub fn move_workspace_entry(
    workspace: State<'_, ApprovedWorkspace>,
    relative_path: String,
    destination_directory: String,
) -> Result<String, String> {
    dirs::move_entry(&workspace.require()?, &relative_path, &destination_directory)
}

#[tauri::command]
pub fn delete_workspace_entry(
    workspace: State<'_, ApprovedWorkspace>,
    relative_path: String,
) -> Result<(), String> {
    dirs::delete_entry(&workspace.require()?, &relative_path)
}

#[tauri::command]
pub fn preview_workspace_deletion(
    workspace: State<'_, ApprovedWorkspace>,
    relative_path: String,
) -> Result<DeletionPreview, String> {
    dirs::preview_deletion(&workspace.require()?, &relative_path)
}
